[Docs] Update ISP Flashing guide (#8149)
qmk/qmk_firmware: docs/isp_flashing_guide.md

# ISP Flashing Guide

> ISP Flashing (also known as ICSP flashing) is the process of programming the microcontroller directly.  This allows you to replace the bootloader, or change the "fuses" on the controller (fuses control a number of hardware and software related functions, such as the speed of the controller, how it boots, and other options).  

Reviewer: ```suggestion
ISP flashing (also known as ICSP flashing) is the process of programming a microcontroller directly.  This allows you to replace the bootloader, or change the "fuses" on the controller, which control a number of hardware- and software-related functions, such as the speed of the controller, how it boots, and other options.  
```

    Memory write error, use debug for more info.
    commands.c:360: Error writing memory data. (err -4)

Or, if you see this sort of message for a Pro Micro based controller:

    avrdude: butterfly_recv(): programmer is not responding
    avrdude: butterfly_recv(): programmer is not responding
    avrdude: verification error, first mismatch at byte 0x002a
            0x2b != 0x75
    avrdude: verification error; content mismatch
    avrdude: verification error; content mismatch

You're likely going to need to ISP flash your board/device to get it working again. 

## Hardware Needed

> You'll need one of the following to actually perform the ISP Flashing (followed by the protocol they use):

Reviewer: ```suggestion
You'll need one of the following to actually perform the ISP flashing (followed by the protocol they use):
```

* [SparkFun PocketAVR](https://www.sparkfun.com/products/9825) - (USB Tiny)
* [USBtinyISP AVR Programmer Kit](https://www.adafruit.com/product/46) - (USB Tiny)
* [Teensy 2.0](https://www.pjrc.com/store/teensy.html) - (avrisp)
* [Pro Micro](https://www.sparkfun.com/products/12640)  - (avrisp)
* [Bus Pirate](https://www.adafruit.com/product/237) - (buspirate)

There are other devices that can be used to ISP flash, but these are the main ones.  Also, all product links are to the official versions. You can source them elsewhere. 

> You'll also need something to wire your "ISP Programmer" to the device that you're programming.  Some PCBs may have ISP headers that you can use directly, but this often isn't the case, so you'll likely need to solder to the controller itself or to different switchs or other components. 

Reviewer 2: ```suggestion
You'll also need something to wire your "ISP Programmer" to the device that you're programming.  Some PCBs may have ISP headers that you can use directly, but this often isn't the case, so you'll likely need to solder to the controller itself or to different switches or other components. 
```

    Pro Micro VCC      <-> Keyboard VCC
    Pro Micro GND      <-> Keyboard GND

!> Note that the 10/B6 pin on the Pro Micro is wired to the RESET/RST pin on the keyboard's controller.  ***DO NOT*** wire the RESET pin on the Pro Micro to the RESET on the keyboard. 

## Flashing your keyboard 

After you have your ISP Programmer set up, and wired to your keyboard, it's time to flash your keyboard.  

### The Bootloader File

> The simplest and quickest way to get thigns back to normal is to flash only a bootloader to the keyboard. Once this is done, you can connect the keyboard normally and flash the keyboard like you normally would. 

Reviewer 2: ```suggestion
The simplest and quickest way to get things back to normal is to flash only a bootloader to the keyboard. Once this is done, you can connect the keyboard normally and flash the keyboard like you normally would. 
```

You can find the stock bootloaders in the [`util/` folder](https://github.com/qmk/qmk_firmware/tree/master/util). Be sure to flash the correct bootloader for your chip:

* [`atmega32u4`](https://github.com/qmk/qmk_firmware/blob/master/util/bootloader_atmega32u4_1_0_0.hex) - Most keyboards, Planck Rev 1-5, Preonic Rev 1-2

> 8 [`Pro Micro`](https://github.com/adafruit/Caterina-Bootloader/blob/master/Built%20Firmwares/Caterina-Leonardo.hex) - The default bootloader for Pro Micro controllers

Reviewer 2: ```suggestion
* [`Pro Micro`](https://github.com/adafruit/Caterina-Bootloader/blob/master/Built%20Firmwares/Caterina-Leonardo.hex) - The default bootloader for Pro Micro controllers
```

Author: ```suggestion
* [`Pro Micro`](https://github.com/sparkfun/Arduino_Boards/blob/master/sparkfun/avr/bootloaders/caterina/Caterina-promicro16.hex) - The default bootloader for Pro Micro controllers
```

### Production Techniques

If you'd like to flash both the bootloader **and** the regular firmware at the same time, there are two options to do so.  Manually, or with the `:production` target when compiling. 

To do this manually: 

1. Open the original firmware .hex file in a text editor
2. Remove the last line (which should be `:00000001FF` - this is an EOF message)
3. Copy the entire bootloader's contents onto a new line (with no empty lines between) and paste it at the end of the original file
4. Save it as a new file by naming it `<keyboard>_<keymap>_production.hex`

?> It's possible to use other bootloaders here in the same way, but __you need a bootloader__, otherwise you'll have to use ISP again to write new firmware to your keyboard.

To do this the easy way, you can flash the board using the `:production` target when compiling.  This compiles the firmware, then compiles the QMK DFU bootloader, and then creates a combined image.  Once this is done, you'll see three files: 
* `<keyboard>_<keymap>.hex`
* `<keyboard>_<keymap>_bootloader.hex`
* `<keyboard>_<keymap>_production.hex`

> The QMK DFU bootloader has only really been tested on `atmega32u4` controllers (such as the Planck boards, and the Pro Micro), and hasn't been tested on other controllers.  However, it will not work on V-USB controllers, such as the `atmega32a` or `atmega32p`. 

Reviewer 2: ```suggestion
The QMK DFU bootloader has only really been tested on `atmega32u4` controllers (such as the Planck boards, and the Pro Micro), and hasn't been tested on other controllers.  However, it will not work on V-USB controllers, such as the `atmega32a` or `atmega328p`. 
```

> The QMK DFU bootloader has only really been tested on `atmega32u4` controllers (such as the Planck boards, and the Pro Micro), and hasn't been tested on other controllers.  However, it will not work on V-USB controllers, such as the `atmega32a` or `atmega328p`. 

Reviewer 3: Shall we highlight this with a `!>`?

Reviewer 3: We should probably change this to "such as the AVR-based Planck boards", too.

You can flash either the bootloader or the production firmware file. The production firmware file will take a lot longer to flash, since it's flashing a lot more data. 

?> Note: You should stay with the same bootloader. If you're using DFU already, switching to QMK DFU is fine. But flashing QMK DFU onto a Pro Micro, for instance, has additional steps needed.

## Flashing Your Bootloader/Production File

Make sure your keyboard is unplugged from any device, and plug in your ISP Programmer.

If you want to change bootloader types, You'll need to use the command line. 

### QMK Toolbox

> 1. `AVRISP device connected` or `USB Ttiny device connected` will show up in yellow

Reviewer 2: I imagine this is a typo.

```suggestion
1. `AVRISP device connected` or `USB Tiny device connected` will show up in yellow
```

Which means everything should be ok! Your board may restart automatically, otherwise, unplug your Teensy and plug in your keyboard - you can leave your Teensy wired to your keyboard while testing things, but it's recommended that you desolder it/remove the wiring once you're sure everything works.

If you're using a SparkFun PocketAVR Programmer, or another USB Tiny based ISP programmer, you will want to use something like this: 

>     avrdude -c usbbiny -P usb -p atmega32u4

Reviewer 2: Typo again?

```suggestion
    avrdude -c usbtiny -P usb -p atmega32u4
```

Author: That's what I get for doing this at 2am

#### Advanced: Changing Fuses

If you're switching bootloaders, such as flashing QMK DFU on a Pro Micro, you will need to change the fuses, in additional to flashing the bootloader hex file.  This is because `caterina` (the Pro Micro bootloader) and `dfu` handle the startup routines differently, and that behavior is controlled by the fuses.  

> !> This is one area that it is very important to be careful, as changing fuses is one of the ways that you can perminently brick your controller.  

Reviewer 2: ```suggestion
!> This is one area that it is very important to be careful, as changing fuses is one of the ways that you can permanently brick your controller.  
```

For this, we are assuming the 5V 16MHz versions of the `atmega32u4` (such as the 5V Pro Micro).

For DFU on the `atmega32u4`, these are the fuse settings that you want: 

> |----------|------------------|

Reviewer 2: Need a table header, or the docs won't render this correctly.

```suggestion
| Fuse     | Setting          |
|----------|------------------|
```

| Low      | `0x5E`           |
| High     | `0xD9` or `0x99` |
| Extended | `0xC3`           |

The High fuse can be 0xD9 or 0x99. The difference is that 0xD9 disables JTAG, which QMK Firmware disables via software as well, while 0x99 doesn't disable JTAG. 

To set this add `-U lfuse:w:0x5E:m -U hfuse:w:0xD9:m -U efuse:w:0xC3:m` to your command.  So the final command should look something like: 

    avrdude -c avrisp -P COM3 -p atmega32u4 -U flash:w:main.hex:i -U lfuse:w:0x5E:m -U hfuse:w:0xD9:m -U efuse:w:0xC3:m

For Caterina on the `atmega32u4`, these are the fuse settings that you want: 

> |----------|--------|

Reviewer 2: ```suggestion
| Fuse     | Setting|
|----------|--------|
```

| Low      | `0xFF` |
| High     | `0xD9` |
| Extended | `0xC3` |

To set this add `-U lfuse:w:0xFF:m -U hfuse:w:0xD8:m -U efuse:w:0xC3:m` to your command.  So the final command should look something like: 

    avrdude -c avrisp -P COM3 -p atmega32u4 -U flash:w:main.hex:i -U lfuse:w:0xFF:m -U hfuse:w:0xD8:m -U efuse:w:0xC3:m

> If you are using a different controller or want different configuration, you can use [this AVR Fuse Calculator](http://www.engbedded.com/fusecalc/) to find a better value for you.

Reviewer 2: This fuse calculator doesn't seem to give the same values for 32u4 DFU as you gave in the table above, but [this one does](http://eleccelerator.com/fusecalc/fusecalc.php).

Please advise.

Author: Default has Extended as 0xF3, but seems to actually default to 0xC3, which is why I picked that value.  

I believe that the defaults that the link uses is what docs say.  However, IIRC, the actual difference matter in practice.

Reviewer: > * Note that some numerical values refer to fuses containing undefined bits (set to '1' here). Depending on the target device these fuse bits will be read either as '0' or '1'. Verification errors will occur if the values are read back with undefined bits set to '0'. Everything is fine if the values read from the device are either the same as programmed, or the following values (undefined set to '0'): Extended: 0x03.

Author: @fauxpark So, would the recommended be `0x03` for the extended fuse, then?  Or?

Reviewer: I just tried setting the efuse to F3 and 03, either way avrdude complains it read back C3. So that is probably the "true" default.
